client/wfprev-war/src/main/angular/src/app/services/spatial-services.ts

```diff
@@ -0,0 +1,393 @@
+            return null;
+        } else geometry = this.stripAltitude(geometry);
+
+        switch (geometry.type) {
+            case 'Point':
+            case 'LineString':
+            case 'Polygon':
+            case 'MultiPoint':
+            case 'MultiLineString':
+            case 'MultiPolygon':
+                return geometry.coordinates;
+            case 'GeometryCollection':
+                return geometry.geometries
+                    .map(geom => this.getCoordinatesFromGeometry(geom))
+                    .find(coords => coords !== null) || null;
+            default:
+                return null;
+        }
+    }
+
+    stripAltitude(geometry: Geometry): Geometry {
+        const strip = (coord: any): any => {
+            if (Array.isArray(coord)) {
+                if (typeof coord[0] === 'number') {
+                    // [lng, lat, alt?] → [lng, lat]
+                    return coord.slice(0, 2);
+                } else {
+                    return coord.map(strip);
+                }
+            }
+            return coord;
+        };
+
+        switch (geometry.type) {
+            case 'Point':
+            case 'MultiPoint':
+            case 'LineString':
+            case 'MultiLineString':
+            case 'Polygon':
+            case 'MultiPolygon':
+                return {
+                    type: geometry.type,
+                    coordinates: strip(geometry.coordinates),
+                    ...(geometry.bbox ? { bbox: geometry.bbox } : {})
+                };
+
+            case 'GeometryCollection':
+                return {
+                    type: 'GeometryCollection',
+                    geometries: geometry.geometries.map(g => this.stripAltitude(g)),
+                    ...(geometry.bbox ? { bbox: geometry.bbox } : {})
+                };
+
+            default:
+                return geometry;
+        }
+    }
+
+    extractGDBGeometry(file: File): Observable<any> {
+        // this will be replaced with node js server in AWS - WFPREV-402
+        const url = 'http://localhost:3000/upload';
+
+        const formData = new FormData();
+        formData.append('file', file);
+
+        return this.httpClient.post<any>(url, formData).pipe(
+            map((response) => {
+                return response.map((geom: any) => {
+                    return this.stripAltitude(geom);
+                });
+            }),
+            catchError((error) => {
+                console.error("Error extracting geodatabase geometry", error);
+                return throwError(() => new Error("Failed to extract geodatabase geometry"));
+            })
+        );
+    }
+
+    public async extractCoordinates(file: File): Promise<number[][]> {
+        const fileType = file.name.split('.').pop()?.toLowerCase();
+        if (!fileType) return [];
+
+        try {
+            if (fileType === 'kml') {
+                return this.extractKMLCoordinates(await file.text()) as unknown as number[][];
```

Comment:
It bypasses type safety (unknown as). can wechange the return type of the method to match what you are actually returning? Position[][][], not number[][],

we can change above method 
public async extractCoordinates(file: File): Promise<number[][]>
to
public async extractCoordinates(file: File): Promise<Position[][][]>

then update the return lines to 

if (fileType === 'kml') {
  return this.extractKMLCoordinates(await file.text());
}

if (['zip', 'gdb'].includes(fileType)) {
  return await this.handleCompressedFile(file);
}

Reply:
Updated, thanks